### PR TITLE
gitserver: Improve handling of failed commit errors

### PR DESCRIPTION
**Pull request by eseliger**
This PR improves how we handle failed to apply patch errors by doing the following:
- Make it a 4xx error so that we don't retry this request through the internalClient
- Make a patch does not apply a fatal error so that the dbworker doesn't retry it
- Actually return the well formatted CreateCommitFromPatch from the client

## Test plan

Verified errors now show up nicer and we don't retry failed patch errors.

### CODE REVIEW

File: internal/gitserver/client.go [hunk 1]
```diff
@@ -1282,22 +1283,13 @@ func (c *clientImplementor) CreateCommitFromPatch(ctx context.Context, req proto
...
-	}
-
 	var res protocol.CreateCommitFromPatchResponse
 	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
 		c.logger.Warn("decoding gitserver create-commit-from-patch response", sglog.Error(err))
 		return "", &url.Error{
 			URL: resp.Request.URL.String(),
 			Op:  "CreateCommitFromPatch",
-			Err: errors.Errorf("CreateCommitFromPatch: http status %d, %v", resp.StatusCode, err),
+			Err: errors.Errorf("CreateCommitFromPatch: http status %d, %s", resp.StatusCode, readResponseBody(resp.Body)),
```
Comment (keegancsmith) [inline internal/gitserver/client.go:1292]: you have already started reading from resp.Body so it is unknown what is left in resp.Body from what json.NewDecoder.Decode has done. Rather read resp.Body into a bytes.Buffer first.

File: internal/httpcli/client.go [hunk 2]
```diff
@@ -552,7 +552,7 @@ func NewRetryPolicy(max int) rehttp.RetryFn {
 			return true
 		}
 
-		if status == 0 || status == 429 || (status >= 500 && status != 501) {
+		if status == 0 || status == http.StatusTooManyRequests || (status >= http.StatusInternalServerError && status != http.StatusNotImplemented) {
```
Comment (keegancsmith) [inline internal/httpcli/client.go:555]: This update looks good. However, in particular I think the replacement of 500 with the constant makes this harder to understand. It was easier to reason about when in the old code when it was clear you are dealing with all 5xx errors.

**Review by keegancsmith — APPROVED** [review]

File: enterprise/internal/batches/reconciler/executor.go [hunk 3]
```diff
@@ -565,6 +565,10 @@ func (e *executor) pushCommit(ctx context.Context, opts protocol.CreateCommitFro
 	if err != nil {
 		var e *protocol.CreateCommitFromPatchError
 		if errors.As(err, &e) {
+			// Make patch does not apply error a fatal error, retries won't help here.
```
Comment (ryanslade) [inline enterprise/internal/batches/reconciler/executor.go:568]: I think this comment needs to be reworded a bit, it's difficult to follow as is.

File: internal/gitserver/client.go [hunk 4]
```diff
@@ -25,10 +25,11 @@ import (
 	"github.com/opentracing/opentracing-go/log"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promauto"
-	"github.com/sourcegraph/sourcegraph/internal/actor"
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/sync/semaphore"
 
+	"github.com/sourcegraph/sourcegraph/internal/actor"
```
Comment (ryanslade) [inline internal/gitserver/client.go:31]: nit: Can you group this with the other sourcegraph imports lower down

**Review by ryanslade — APPROVED** [review]

**Pull request merged**